### PR TITLE
EWB-2500: Support protection equipment

### PR DESCRIPTION
**Pull request by Lax125**
# Description

Implements model change to support protection equipment. See https://app.clickup.com/t/6929263/EWB-2473 for the spec.

# Associated tasks:

Tasks blocking this one:
- [x] Merge https://github.com/zepben/evolve-sdk-python/pull/122

Tasks this is blocking:
- https://github.com/zepben/pp-translator/pull/20
- https://github.com/zepben/edith-sdk/pull/4
- https://github.com/zepben/ewb-sdk-examples-python/pull/3

# Checklist:

- [x] I have performed a self review of my own code.
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have updated the changelog.
- [x] I have commented my code in any hard-to-understand or hacky areas.
- [x] I have handled all new warnings generated by the compiler or IDE.
- [x] I have considered if this is a breaking change and have communicated it with other team members if so.

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-2473 Data model change spec](https://app.clickup.com/t/6929263/EWB-2473)

**Comment by ryanjcoleman** [comment]
Task linked: [EWB-2500 Update Python SDK](https://app.clickup.com/t/6929263/EWB-2500)

File: src/zepben/evolve/model/cim/iec61970/base/protection/current_relay.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    inverse_time_flag: Optional[bool] = None
+    """Set true if the current relay has inverse time characteristic."""
+
+    time_delay_1: Optional[float] = None
+    """Inverse time delay number 1 for current limit number 1 in seconds."""
+
+    @property
+    def current_relay_info(self) -> Optional[CurrentRelayInfo]:
+        """Datasheet information for this :class:`CurrentRelay`."""
+        return self.asset_info
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/current_relay.py:29]: This is a type strengthening property, like `ShuntCompensator.shunt_compensator_info`.

[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/current_relay.py:29]: @sonatype-lift ignore

File: src/zepben/evolve/model/cim/iec61970/base/protection/protection_equipment.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    relay_delay_time: Optional[float] = None
+    """The time delay from detection of abnormal conditions to relay operation in seconds."""
+
+    protection_kind: ProtectionKind = ProtectionKind.UNKNOWN
+    """The kind of protection being provided by this protection equipment."""
+
+    _protected_switches: Optional[List[ProtectedSwitch]] = None
+
+    def __init__(self, protected_switches: Iterable[ProtectedSwitch] = None, **kwargs):
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/protection/protection_equipment.py:35]: For whatever reason, `Optional` is omitted in `__init__` for optional parameters in this repo. On one hand it discourages manually setting the param to `None` which has no effect, but on the other hand it makes it less clear that the parameter is optional. @kgreav your thoughts?

File: src/zepben/evolve/services/common/reference_resolvers.py [hunk 3]
```diff
@@ -174,10 +188,14 @@ def _resolve_diag_diagobj(diag, diag_obj):
 
 conductor_to_wire_info_resolver = ReferenceResolver(Conductor, WireInfo, lambda t, r: setattr(t, 'asset_info', r))
 
+current_relay_to_current_relay_info_resolver = ReferenceResolver(CurrentRelay, CurrentRelayInfo, lambda t, r: setattr(t, 'asset_info', r))
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/services/common/reference_resolvers.py:191]: Is a `dataclassy.dataclass`

[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/services/common/reference_resolvers.py:191]: @sonatype-lift ignore

File: src/zepben/evolve/model/cim/iec61970/base/wires/switch.py [hunk 4]
```diff
@@ -42,13 +46,29 @@ class Switch(ConductingEquipment):
...
     """Tells if the switch is considered open when used as input to topology processing."""
 
     _normally_open: int = 0
     """The attribute is used in cases when no Measurement for the status value is present. If the Switch has a status measurement the Discrete.normalValue 
     is expected to match with the Switch.normalOpen."""
 
+    @property
+    def switch_info(self) -> Optional[SwitchInfo]:
+        """Datasheet information for this Switch."""
+        return self.asset_info
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/wires/switch.py:62]: This is a type strengthening property, like `ShuntCompensator.shunt_compensator_info`.

[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/model/cim/iec61970/base/wires/switch.py:62]: @sonatype-lift ignore

File: src/zepben/evolve/database/sqlite/tables/associations/protection_equipment_protected_switches.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+from typing import List
+
+from zepben.evolve.database.sqlite.tables.column import Column, Nullable
+from zepben.evolve.database.sqlite.tables.sqlite_table import SqliteTable
+
+__all__ = ["TableProtectionEquipmentProtectedSwitches"]
+
+
+class TableProtectionEquipmentProtectedSwitches(SqliteTable):
+    protection_equipment_mrid: Column = None
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/database/sqlite/tables/associations/protection_equipment_protected_switches.py:15]: `dataclassy` needs a default value to use as a factory.

[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/database/sqlite/tables/associations/protection_equipment_protected_switches.py:15]: @sonatype-lift ignore

**Comment by Lax125** [comment]
Sorry for the late update, some minor fixes to some test names along with adding missing `await`s.

**Review by kgreav — CHANGES_REQUESTED** [review]
Nice work, only things I can see you've missed is updating `test_network_service_comparator.py` and the changelog update.

File: changelog.md [hunk 6]
```diff
@@ -20,6 +20,16 @@
...
+  * `Disconnector` moved to `disconnector`
+  * `Fuse` moved to `fuse`
+  * `Jumper` moved to `jumper`
+  * `LoadBreakSwitch` moved to `load_break_switch`
+  * `ProtectedSwitch` moved to `protected_switch`
+  * `Recloser` moved to `recloser`
+  
+  Note that `from zepben.evolve import <ClassName>` will still work as usual for all of the above classes.
 
 ### New Features
```
[the remark this answers is not in the document]
Comment (Lax125) [reply changelog.md:35]: Added notes.

File: src/zepben/evolve/database/sqlite/readers/network_cim_reader.py [hunk 7]
```diff
@@ -105,6 +107,13 @@ def load_shunt_compensator_info(self, table: TableShuntCompensatorInfo, rs: Resu
 
         return self._load_asset_info(shunt_compensator_info, table, rs) and self._add_or_throw(shunt_compensator_info)
 
+    def load_switch_info(self, table: TableSwitchInfo, rs: ResultSet, set_last_mrid: Callable[[str], str]) -> bool:
+        switch_info = SwitchInfo(mrid=set_last_mrid(rs.get_string(table.mrid.query_index)))
```
[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/database/sqlite/readers/network_cim_reader.py:112]: `get_string` raises by default

[the remark this answers is not in the document]
Comment (Lax125) [reply src/zepben/evolve/database/sqlite/readers/network_cim_reader.py:112]: @sonatype-lift ignore

**Review by kgreav — APPROVED** [review]

**Pull request merged**